src/command.rs

```diff
@@ -428,30 +428,226 @@ pub trait Parameters: Sized {
 }
 
+pub trait BrowserCapabilities {
+    fn init(&mut self, &BTreeMap<String, Json>);
+    fn browser_name(&mut self, &BTreeMap<String, Json>) -> Option<String>;
+    fn browser_version(&mut self, &BTreeMap<String, Json>) -> Option<String>;
+    fn platform_name(&mut self, &BTreeMap<String, Json>) -> Option<String>;
+    fn platform_version(&mut self, &BTreeMap<String, Json>) -> Option<String>;
+    fn accept_insecure_certs(&mut self, &BTreeMap<String, Json>) -> bool;
+    fn compare_browser_version(&mut self, &str, &str) -> WebDriverResult<bool>;
+}
+
+pub trait CapabilitiesMatching {
+    fn match_browser<T: BrowserCapabilities>(&self, browser_capabilities: &mut T)
+                                             -> Option<BTreeMap<String, Json>>;
+}
+
 #[derive(PartialEq)]
-pub struct NewSessionParameters {
-    pub desired: BTreeMap<String, Json>,
-    pub required: BTreeMap<String, Json>,
+pub enum NewSessionParameters {
+    Spec(SpecNewSessionParameters),
+    Legacy(LegacyNewSessionParameters)
 }
 
-impl NewSessionParameters {
-    pub fn get(&self, name: &str) -> Option<&Json> {
-        self.required.get(name).or_else(|| self.desired.get(name))
+impl Parameters for NewSessionParameters {
+    fn from_json(body: &Json) -> WebDriverResult<NewSessionParameters> {
+        let data = try_opt!(body.as_object(),
+                            ErrorStatus::UnknownError,
+                            "Message body was not an object");
+        if data.get("capabilities").is_some() {
+            Ok(NewSessionParameters::Spec(try!(SpecNewSessionParameters::from_json(body))))
+        } else {
+            Ok(NewSessionParameters::Legacy(try!(LegacyNewSessionParameters::from_json(body))))
+        }
     }
+}
 
-    pub fn consume(&mut self, name: &str) -> Option<Json> {
-        let required = self.required.remove(name);
-        let desired = self.desired.remove(name);
-        if required.is_some() {
-            required
-        } else {
-            desired
+impl ToJson for NewSessionParameters {
+    fn to_json(&self) -> Json {
+        match self {
+            &NewSessionParameters::Spec(ref x) => x.to_json(),
+            &NewSessionParameters::Legacy(ref x) => x.to_json()
         }
     }
 }
 
-impl Parameters for NewSessionParameters {
-    fn from_json(body: &Json) -> WebDriverResult<NewSessionParameters> {
+impl CapabilitiesMatching for NewSessionParameters {
+    fn match_browser<T: BrowserCapabilities>(&self, browser_capabilities: &mut T)
+                                             -> Option<BTreeMap<String, Json>> {
+        match self {
+            &NewSessionParameters::Spec(ref x) => x.match_browser(browser_capabilities),
+            &NewSessionParameters::Legacy(ref x) => x.match_browser(browser_capabilities)
+        }
+    }
+}
+
+#[derive(PartialEq)]
+pub struct SpecNewSessionParameters {
+    pub alwaysMatch: BTreeMap<String, Json>,
+    pub firstMatch: Vec<BTreeMap<String, Json>>,
```

Comment:
We are not 100% consistent on this but it does make things easier if the field names match the wire protocol since we will be able to use `serde` directly in the future.